assets/js/base/components/shipping-rates-control/packages.js

```diff
@@ -12,23 +14,29 @@ import './style.scss';
 const Packages = ( {
 	className,
 	noResultsMessage,
-	onChange = () => {},
 	renderOption,
-	selected = [],
-	shippingRates,
+	shippingRates = [],
```

Comment:
well this is two folds, if we're going to go with the approach you’re suggesting, we will need to lift all the way up to `FullCart`, I don’t want this, I want to isolate hooks to appropriate components to avoid unnecessary full rerender (only packages need to rerender, not the full cart), having all hooks and all logic there will cause performance issues along the way, I would also keep shipping rates within this component if I can, but since we have preview data that is not coming from a hook, I have to pass it to full cart.

Reply:
I can move those to the top until we find a solution for mocking data store in the Editor so that we can rid of all top level hook usage.

Reply:
> well this is two folds, if we're going to go with the approach you’re suggesting, we will need to lift all the way up to FullCart,

Why does it have to go all the way to `FullCart`?  I'm suggesting you lift state up to a _hook_.  You can still pull in the `useShippingRates` hook here (or in `<ShippingRatesControl />` - which is where I was thinking it would be added).

>  I want to isolate hooks to appropriate components to avoid unnecessary full rerender (only packages need to rerender, not the full cart), having all hooks and all logic there will cause performance issues along the way

I'm wary of assertive statements like "will cause performance issues along the way", unless it's very clear an approach will cause that.  All hooks and logic don't need to get implemented in the `FullCart` and then all props get pushed down.

> but since we have preview data that is not coming from a hook, I have to pass it to full cart.

I'm not sure I understand the relevance here? Why not just pass along an `isEditor` prop from the `<Editor/>` component to full cart and that can be used to determine whether shipping rates come from the preview data or not?

Reply:
Personally I'd avoid having mixing preview handling in on this pull as it could block it from getting merged unnecessarily.

Reply:
assuming we’re getting shipping rates using`useShippingRates`, we will have the mixed preview data thing, I think we can the `useSelectShippingRates` call in packages (the nearest place they’re used in), and refactor in the future?

Reply:
>I'm wary of assertive statements like "will cause performance issues along the way", unless it's very clear an approach will cause that. All hooks and logic don't need to get implemented in the FullCart and then all props get pushed down.

While there is no direct implications that are obvious, having a lot of logic in the root is going to be problematic for slower devices, as a rule of thumb, we should always try to avoid unnecessary rerenders (our Cart block rerenders 3 or 4 times on initial load only)

Reply:
> we should always try to avoid unnecessary rerenders 

I agree, but premature optimization can be expensive too.  Re-renders aren't necessarily bad, it's _expensive_ re-renders that are bad. So let's avoid premature optimization unless it's clear expensive re-rendering is happening.